### PR TITLE
New falling leaf script based on the existing rain.js

### PR DESCRIPTION
**Pull request by BingShearer**
Pushing falling leaves script for Entry to codebase as requested by Chris.

### CODE REVIEW

**Comment by PhilipRosedale** [comment]
Nice!  Show this to me monday.

**Comment by BingShearer** [comment]
Note: Assignment Client Bug will not allow "complexMovement" to function correctly on servers for the time being, not a script issue. Would still be nice to have this in the codebase.

File: examples/leaves.js [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    leafSize: { x: 0.3, y: 0.00, z: .3},
+    leafFallSpeed: .4,
+    leafLifetime: 100,
+    leafSpinMax: 30,
+    debug: false,
+    maxLeaves: 100,
+    leafDeleteOnTearDown: true,
+    complexMovement: true,
+    floorHeight: 2143.5,
+    windFactor: .5,
```
Comment (AndrewMeadows) [inline examples/leaves.js]: Use **0.5** instead of **.5** here, and similarly for the **.3** at line 277.

File: examples/leaves.js [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        leafTimer = Script.setInterval(function () {
+            if (leafCount <= maxLeaves - 1) {
+                createleaf()
+            }
+        }, 60000 / leavesPerMinute);
+    }
+    Script.setInterval(function () {
+        nearbyEntities = Entities.findEntities(squallOrigin, squallRadius);
+            newLeafMovement()
+    }, 100);
```
Comment (AndrewMeadows) [inline examples/leaves.js]: Bad indentation at line 178.

File: examples/leaves.js [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                        leafLocalHorVec = Vec3.cross(leafYinWFVec, yVec),
+                        leafMostDownVec = Vec3.cross(leafYinWFVec, leafLocalHorVec),
+                        leafDesiredVel = Vec3.multiply(leafMostDownVec, windFactor);    
+                    Entities.editEntity(currentLeaf, {
+                        angularVelocity: randomRotationSpeed,
+                        velocity: {
+                            x: (leafDesiredVel.x - currentLeafVelocity.x) * windFactor + currentLeafVelocity.x,
+                            y: (leafDesiredVel.y - currentLeafVelocity.y) * windFactor + currentLeafVelocity.y,
+                            z: (leafDesiredVel.z - currentLeafVelocity.z) * windFactor + currentLeafVelocity.z
+                        }
```
Comment (AndrewMeadows) [inline examples/leaves.js]: We have a Vec3 object that makes it easier to do Vec3 operations on JS objects if you want to use them.  So the calculation of new velocity could look something like this: 

    var deltaVelocity = Vec3.subtract(leafDesiredVel, currentLeafVelocity);
    var newVelocity = Vec3.sum(currentLeafVelocity, Vec3.multiply(windFactor, deltaVelocity));

File: examples/leaves.js [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    function createleaf() {
+        var angle,
+            radius,
+            offset,
+            leaf,
+            spin = { x: 0, y: 0, z: 0 },           
+            i;
+
+        // HACK: Work around leaves not always getting velocities set at creation
+        for (i = 0; i < leaves.length; i += 1) {
```
Comment (ZappoMan) [inline examples/leaves.js]: why not i++?

File: examples/leaves.js [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+            Entities.editEntity(leaves[i], leafVelocities[i]);
+        }
+
+        angle = Math.random() * 10;
```
Comment (ZappoMan) [inline examples/leaves.js]: coding standard: no magic numbers... why 10? give it a name.

File: examples/leaves.js [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+        offset = Vec3.multiplyQbyV(Quat.fromPitchYawRollDegrees(0, angle, 0), { x: 0, y: -0.1, z: radius });
+        leafProperties.position = Vec3.sum(squallOrigin, offset);
+        if (properties.leafSpinMax > 0 && !complexMovement) {
+            spin = {
+                x: Math.random() * maxSpinRadians,
+                y: Math.random() * maxSpinRadians,
+                z: Math.random() * maxSpinRadians
+            };
+            leafProperties.angularVelocity = spin;
+        }
```
Comment (ZappoMan) [inline examples/leaves.js]: coding standard: brace placement...

File: examples/leaves.js [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+                        leafLocalHorVec = Vec3.cross(leafYinWFVec, yVec),
+                        leafMostDownVec = Vec3.cross(leafYinWFVec, leafLocalHorVec),
+                        leafDesiredVel = Vec3.multiply(leafMostDownVec, windFactor),
+                        leafVelDelt = Vec3.subtract(leafDesiredVel, leafCurrentVel),
+                        leafNewVel = Vec3.sum(leafCurrentVel, Vec3.multiply(leafVelDelt, windFactor));
+                    Entities.editEntity(currentLeaf, {
+                        angularVelocity: randomRotationSpeed,
+                        velocity: leafNewVel
+                    })
+                }
```
Comment (ZappoMan) [inline examples/leaves.js]: coding standard: brace placement...

File: examples/leaves.js [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+                else if (leafHeight <= floorHeight) {
+                    if (!leafDeleteOnGround) {
+                        Entities.editEntity(nearbyEntities[i], {
+                            locked: false,
+                            velocity: { x: 0, y: 0, z: 0 },
+                            angularVelocity: { x: 0, y: 0, z: 0 }
+                        })
+                    }
```
Comment (ZappoMan) [inline examples/leaves.js]: coding standard: brace placement...

File: examples/leaves.js [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        }
+    }
+
+
+
+    processProperties();
+    setUp();
+    Script.scriptEnding.connect(tearDown);
+
+    return {
```
Comment (ZappoMan) [inline examples/leaves.js]: why use this construct? why not just let the function end?

File: examples/leaves.js (hunk 4 above)
Comment (howard-stearns) [inline examples/leaves.js]: Some JavaScript programmers, including Douglas  Crickford (but not me), consider ++ to be an anti pattern. Others (including me) use linters that reject this. (One can also tell lint to accept plus plus)

Comment (ZappoMan) [inline examples/leaves.js]: Seeing how he uses ++ a couple lines later I'm unconvinced.

File: examples/leaves.js (hunk 9 above)
Comment (BingShearer) [inline examples/leaves.js]: Because this function gives the option to delete the leaves, or debugging entities when the script ends.

**Pull request merged**